### PR TITLE
clientv3: add GoDoc to OpOption

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

File: clientv3/op.go [hunk 1]
```diff
@@ -119,16 +119,26 @@ func (op *Op) applyOpts(opts []OpOption) {
 
 type OpOption func(*Op)
 
+// WithLease attaches a lease ID to a key in Put request.
 func WithLease(leaseID lease.LeaseID) OpOption {
 	return func(op *Op) { op.leaseID = leaseID }
 }
+
+// WithLimit limits the number of results to return from Get or Delete request.
```
Comment (xiang90) [inline clientv3/op.go]: does it limit delete request?

File: clientv3/op.go [hunk 2]
```diff
@@ -119,16 +119,26 @@ func (op *Op) applyOpts(opts []OpOption) {
...
 func WithLimit(n int64) OpOption { return func(op *Op) { op.limit = n } }
+
+// WithRev specifies the store revision for Get or Delete request.
 func WithRev(rev int64) OpOption { return func(op *Op) { op.rev = rev } }
```
Comment (xiang90) [inline clientv3/op.go]: this does not apply to delete i think.

File: clientv3/op.go (hunk 1 above)
Comment (gyuho) [inline clientv3/op.go]: Right. Delete request ignores this. Will fix. Thanks.

File: clientv3/op.go [hunk 3]
```diff
@@ -119,16 +119,26 @@ func (op *Op) applyOpts(opts []OpOption) {
...
+
+// WithSort specifies the order in Get results.
```
Comment (xiang90) [inline clientv3/op.go]: WithSort specifies the order in Get requests when WithRange or WithPrefix is also specified.

File: clientv3/op.go [hunk 4]
```diff
@@ -145,10 +155,22 @@ func WithPrefix() OpOption {
...
+// the keys in the range [key, end).
 func WithRange(endKey string) OpOption {
 	return func(op *Op) { op.end = []byte(endKey) }
 }
+
+// WithFromKey specifies the range of Get or Delete requests
+// that keys are greater than or equal to key in the argument.
 func WithFromKey() OpOption { return WithRange("\x00") }
+
+// WithSerializable serializes Get and Delete requests. Which are
```
Comment (xiang90) [inline clientv3/op.go]: this does not apply to delete request. delete request will always go through raft.

File: clientv3/op.go [hunk 5]
```diff
@@ -145,10 +155,22 @@ func WithPrefix() OpOption {
...
+// linearizable by default. Serialized requests are served locally
+// without going through consensus algorithm, useful for achieving
+// lower latency, but this comes with weak-consistency.
```
Comment (xiang90) [inline clientv3/op.go]: this is still strong-consistency. linerizable and consistent are different.

File: clientv3/op.go (hunk 4 above)
Comment (gyuho) [inline clientv3/op.go]: Yeah just fixed. Will fix raft comment part.

File: clientv3/op.go (hunk 2 above)
Comment (gyuho) [inline clientv3/op.go]: Yeah just deleted. Thanks

File: clientv3/op.go [hunk 6]
```diff
@@ -119,16 +119,26 @@ func (op *Op) applyOpts(opts []OpOption) {
 
 type OpOption func(*Op)
```
Comment (xiang90) [inline clientv3/op.go]: doc string OpOption too?

**Comment by heyitsanthony** [comment]
lgtm

**Comment by xiang90** [comment]
defer to @heyitsanthony

**Comment by xiang90** [comment]
@hongchaodeng Can you please also give this a look?

File: clientv3/op.go [hunk 7]
```diff
@@ -145,10 +157,22 @@ func WithPrefix() OpOption {
...
 }
+
+// WithFromKey specifies the range of 'Get' or 'Delete' requests
+// that keys are greater than or equal to key in the argument.
 func WithFromKey() OpOption { return WithRange("\x00") }
+
+// WithSerializable serializes 'Get' requests. Which are
+// linearizable by default. Serialized requests are served locally
+// without going through consensus algorithm, useful for achieving
+// lower latency.
```
Comment (gyuho) [inline clientv3/op.go]: @xiang90 Can you check this again? Thanks.

File: clientv3/op.go [hunk 8]
```diff
@@ -117,18 +117,30 @@ func (op *Op) applyOpts(opts []OpOption) {
...
 	return func(op *Op) { op.leaseID = leaseID }
 }
+
+// WithLimit limits the number of results to return from 'Get' request.
 func WithLimit(n int64) OpOption { return func(op *Op) { op.limit = n } }
+
+// WithRev specifies the store revision for 'Get' request.
 func WithRev(rev int64) OpOption { return func(op *Op) { op.rev = rev } }
+
+// WithSort specifies the order in 'Get' requests
```
Comment (hongchaodeng) [inline clientv3/op.go]: WithSort specifies the ordering in 'Get' request.
It requires WithRange and/or WithPrefix to be specified too.

File: clientv3/op.go [hunk 9]
```diff
@@ -117,18 +117,30 @@ func (op *Op) applyOpts(opts []OpOption) {
...
+// when WithRange or WithPrefix is also specified.
 func WithSort(tgt SortTarget, order SortOrder) OpOption {
```
Comment (hongchaodeng) [inline clientv3/op.go]: What's tgt, order?

File: clientv3/op.go [hunk 10]
```diff
@@ -117,18 +117,30 @@ func (op *Op) applyOpts(opts []OpOption) {
...
 	return func(op *Op) {
 		op.sort = &SortOption{tgt, order}
 	}
 }
+
+// WithPrefix to 'Get' or 'Delete' by prefix.
```
Comment (hongchaodeng) [inline clientv3/op.go]: Can you clarify it?

File: clientv3/op.go (hunk 9 above)
Comment (gyuho) [inline clientv3/op.go]: Will document it as well. Thanks.

File: clientv3/op.go [hunk 11]
```diff
@@ -145,10 +157,22 @@ func WithPrefix() OpOption {
...
 }
+
+// WithRange specifies the range of 'Get' or 'Delete' requests.
+// For example, 'Get' requests with 'WithRange(end)' returns
+// the keys in the range [key, end).
 func WithRange(endKey string) OpOption {
 	return func(op *Op) { op.end = []byte(endKey) }
 }
+
+// WithFromKey specifies the range of 'Get' or 'Delete' requests
```
Comment (hongchaodeng) [inline clientv3/op.go]: WithFromKey specifies the range of 'Get' or 'Delete' requests to be equal and greater than they key in the argument.

File: clientv3/op.go [hunk 12]
```diff
@@ -145,10 +157,22 @@ func WithPrefix() OpOption {
...
+// that keys are greater than or equal to key in the argument.
 func WithFromKey() OpOption { return WithRange("\x00") }
+
+// WithSerializable serializes 'Get' requests. Which are
```
Comment (hongchaodeng) [inline clientv3/op.go]: WithSerializable makes 'Get' request serializable. By default, it's linearizable. Serializable requests are better for lower latency requirement.

**Comment by gyuho** [comment]
@hongchaodeng All addressed. PTAL. Thanks!

File: clientv3/op.go [hunk 13]
```diff
@@ -145,10 +164,21 @@ func WithPrefix() OpOption {
...
+
+// WithRange specifies the range of 'Get' or 'Delete' requests.
+// For example, 'Get' requests with 'WithRange(end)' returns
+// the keys in the range [key, end).
 func WithRange(endKey string) OpOption {
 	return func(op *Op) { op.end = []byte(endKey) }
 }
+
+// WithFromKey specifies the range of 'Get' or 'Delete' requests
+// to be equal and greater than they key in the argument.
```
Comment (hongchaodeng) [inline clientv3/op.go]: Sorry..
"equal or greater"

Comment (gyuho) [inline clientv3/op.go]: Right. Just fixed. PTAL. Thanks.

**Comment by gyuho** [comment]
merging in green lights if there are no more feedback. thanks all.

File: clientv3/op.go [hunk 14]
```diff
@@ -117,18 +117,37 @@ func (op *Op) applyOpts(opts []OpOption) {
...
+// specifies the target to sort by: key, version, revisions, value.
+// 'SortOrder' can be either 'SortNone', 'SortAscend', 'SortDescend'.
 func WithSort(tgt SortTarget, order SortOrder) OpOption {
 	return func(op *Op) {
 		op.sort = &SortOption{tgt, order}
 	}
 }
+
+// WithPrefix enables 'Get' or 'Delete' requests to operate on
+// the keys with matching prefix. For example, 'Get(foo)' can return
```
Comment (hongchaodeng) [inline clientv3/op.go]: Get(foo, WithPrefix())

File: clientv3/op.go [hunk 15]
```diff
@@ -117,18 +117,37 @@ func (op *Op) applyOpts(opts []OpOption) {
 	}
 }
 
+// OpOption configures Op operations.
```
Comment (xiang90) [inline clientv3/op.go]: configures Operations like Get, Put, Delete.

File: clientv3/op.go [hunk 16]
```diff
@@ -117,18 +117,37 @@ func (op *Op) applyOpts(opts []OpOption) {
...
+// WithRev specifies the store revision for 'Get' request.
+//
+// TODO: support Watch API
 func WithRev(rev int64) OpOption { return func(op *Op) { op.rev = rev } }
+
+// WithSort specifies the ordering in 'Get' request. It requires
+// 'WithRange' and/or 'WithPrefix' to be specified too. 'SortTarget'
+// specifies the target to sort by: key, version, revisions, value.
+// 'SortOrder' can be either 'SortNone', 'SortAscend', 'SortDescend'.
 func WithSort(tgt SortTarget, order SortOrder) OpOption {
```
Comment (hongchaodeng) [inline clientv3/op.go]: tgt -> target?

File: clientv3/op.go [hunk 17]
```diff
@@ -117,18 +117,37 @@ func (op *Op) applyOpts(opts []OpOption) {
...
+// WithLimit limits the number of results to return from 'Get' request.
 func WithLimit(n int64) OpOption { return func(op *Op) { op.limit = n } }
+
+// WithRev specifies the store revision for 'Get' request.
+//
+// TODO: support Watch API
 func WithRev(rev int64) OpOption { return func(op *Op) { op.rev = rev } }
+
+// WithSort specifies the ordering in 'Get' request. It requires
+// 'WithRange' and/or 'WithPrefix' to be specified too. 'SortTarget'
```
Comment (hongchaodeng) [inline clientv3/op.go]: SortTarget -> target
SortOrder -> order

File: clientv3/op.go (hunk 15 above)
Comment (gyuho) [inline clientv3/op.go]: just fixed. thanks!

File: clientv3/op.go (hunk 16 above)
Comment (gyuho) [inline clientv3/op.go]: just fixed. thanks!

File: clientv3/op.go (hunk 17 above)
Comment (gyuho) [inline clientv3/op.go]: just fixed. thanks!

**Comment by hongchaodeng** [comment]
LGTM

**Comment by gyuho** [comment]
thanks. will merge after greens.

**Pull request merged**